### PR TITLE
feat: Add support for multiple namespaces

### PR DESCRIPTION
**Pull request by chinhuang007**
Add multi-namespace support so one controller can manage resources
such as ServingRuntimes and Predictors in multiple namespaces

Signed-off-by: Chin Huang <chhuang@us.ibm.com>

#### Motivation
This PR is to add the multiple namespace support as described in #33, and documented in https://docs.google.com/document/d/12Fscl9yEjDrOAqNEyRhA23MlqUYiHSDOP6XNi5lQYGA/edit#heading=h.6oqyeffyxm7q

#### Modifications
In phase one, we will make the controller to work with multiple namespaces while leave ServingRuntimes as namespaced. The changes are
- controllers to keep track of multiple namespaces
- resolver to work with multiple namespaces
- modelmesh to take namespace as part of service endpoints
- rbac to be cluster scope
- lifecycle support for service, etcd secret, and internal configmap
- install scripts and docs

#### Result
User can install modelmesh-serving once in a controller namespace and deploy ServingRuntimes and Predictors into multiple user namespaces

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -13,14 +13,16 @@
 # limitations under the License.
 ---
 apiVersion: rbac.authorization.k8s.io/v1
-kind: Role
+kind: ClusterRole
```
Comment (njhill) [inline config/rbac/role.yaml:16]: It would be good to move all the RBAC changes into a separate PR. This will help to make sure everything still works fine with the current permissions and single namespace.

**Review by njhill — COMMENTED** [review]
Thanks @chauhang, looks good, made a few comments inline.

File: docs/install/README.md [hunk 2]
```diff
@@ -55,11 +57,14 @@ Please be aware that:
 
 For more details see the [built-in runtime configuration](../configuration/built-in-runtimes.md)
 
-In addition, the following resources will be created in the same namespace:
+In addition, the following resources will be created in the control plane namespace:
 
 - `model-serving-defaults` - ConfigMap holding default values tied to a release, should not be modified. Configuration can be overriden by creating a user ConfigMap, see [configuration](../configuration)
 - `tc-config` - ConfigMap used for some internal coordination
 - `storage-config` - Secret holding config for each of the storage backends from which models can be loaded - see [the example](../predictors/)
+- `model-serving-etcd` - Secret providing access to the Etcd cluster - see [instructions](../install/install-script.md#setup-the-etcd-connection-information)
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply docs/install/README.md:65]: I thought "model-serving-etcd" is similar to "storage-config", up to the user to provide. The line above has "`storage-config` - Secret holding config for each of the storage backends from which models can be loaded - see [the example](../predictors/)". So the word "created" seems to include both by code and user.

File: controllers/servingruntime_controller.go [hunk 3]
```diff
@@ -189,7 +256,9 @@ func (r *ServingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return RequeueResult, fmt.Errorf("could not determine replicas: %w", err)
 	}
 	mmDeployment.Replicas = replicas
-	if err = mmDeployment.Apply(ctx); err != nil {
+	err = mmDeployment.Apply(ctx)
+
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply controllers/servingruntime_controller.go]: Changed back.

File: controllers/servingruntime_controller.go [hunk 4]
```diff
@@ -336,10 +405,11 @@ func (r *ServingRuntimeReconciler) getRuntimesSupportingPredictor(ctx context.Co
 	for i := range runtimes.Items {
 		rt := &runtimes.Items[i]
 		if runtimeSupportsPredictor(rt, p) {
-			srnns = append(srnns, types.NamespacedName{
+			srnn := types.NamespacedName{
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply controllers/servingruntime_controller.go]: Changed back.

File: pkg/mmesh/modelmesh_events.go [hunk 5]
```diff
@@ -123,7 +123,7 @@ func (mes *ModelMeshEventStream) UpdateWatchedService(ctx context.Context,
 				mes.logger.Error(err, "Could not close existing etcd client")
 			}
 		}
-		if err := mes.connectToEtcd(ctx, etcdSecretName); err != nil {
+		if err := mes.connectToEtcd(ctx, etcdSecretName, namespace); err != nil {
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply pkg/mmesh/modelmesh_events.go]: Removed

File: pkg/mmesh/modelmesh_events.go [hunk 6]
```diff
@@ -202,6 +202,10 @@ func (mes *ModelMeshEventStream) refreshWatches(nw *namespaceWatch, namespace, s
 					if hashIdx > ownerIdx && key[hashIdx] == '-' {
 						// Infer predictor/vmodel and source ids from concrete model id by removing hash suffix
 						sourceId, predictorName := key[ownerIdx:hashIdx], key[:ownerIdx-2]
+						i := strings.LastIndex(namespace, "_")
+						if i > 0 {
+							namespace = namespace[i+1:]
+						}
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply pkg/mmesh/modelmesh_events.go:208]: It went into a loop of adding sourceId + "_" without this.

[the remark this answers is not in the document]
Comment (njhill) [reply pkg/mmesh/modelmesh_events.go:208]: Are you sure that's still the case? There was an oversight in the first version of my PR that caused this, but it was fixed before we merged it. Could you try taking out these changes to verify it's still a problem?
If so then there's another bug somewhere we should fix instead of making these changes.

File: docs/install/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (njhill) [reply docs/install/README.md:65]: They are a bit different:
- The etcd secret must be created by the user as a prereq to installation and should be basically static and never (or very rarely) change. However _if_ the `--quickstart` install option is used it will run etcd and create the secret for you.
- The storage config secret does not need to be created by the user, an empty one will be created during installation if it doesn't exist. The user can update this (add/update/delete entries within it) and the system will pick up those changes dynamically. You just need to make sure you add a storage entry to the cert before it can be used by an InferenceService/Predictor. If the `--quickstart` option is used it will also set up minio for you and add an entry for it to the secret.

File: scripts/install.sh [hunk 7]
```diff
@@ -258,4 +264,22 @@ wait_for_pods_ready "-l control-plane=modelmesh-controller"
...
+  kustomize build runtimes --load-restrictor LoadRestrictionsNone > runtimes.yaml
+  cp dependencies/user_ns.yaml .
+  sed -i "s/controller_namespace/${namespace}/g" user_ns.yaml
+
+  for user_ns in "${user_ns_array[@]}"; do
+    if ! kubectl get namespaces $user_ns >/dev/null; then
+      echo "Kube namespace does not exist: $user_ns. Will skip."
+    else 
+      kubectl label namespace ${user_ns} modelmesh-enabled="true"
+      kubectl apply -f user_ns.yaml -n ${user_ns}
```
Comment (njhill) [inline scripts/install.sh:277]: I think we only want to do do this in the `--quickstart` case. Also maybe we should rename it to something like `minio-storage-secret.yaml`?

Comment (chinhuang007) [reply scripts/install.sh]: Sure. Updated.

File: docs/install/README.md [hunk 8]
```diff
@@ -12,9 +12,11 @@
 
 We provide an install script to quickly run ModelMesh Serving with a provisioned etcd server. This may be useful for experimentation or development but should not be used in production.
 
-## Namespace Scope
+## Cluster Scope
 
-ModelMesh Serving is namespace scoped, meaning all of its components must exist within a single namespace and only one instance of ModelMesh Serving can be installed per namespace. Multiple ModelMesh Serving instances can be installed in separate namespaces within the cluster.
+ModelMesh Serving is cluster scoped, meaning its components can exist in multiple user namespaces which are controlled by one instance of ModelMesh Serving Controller in the control plane namespace. Only one ModelMesh Serving Controller instance can be installed within a Kubernetes cluster.
+
+A namespace label `modelmesh-enabled` needs to be "true" to enable a user namespace for ModelMesh Serving.
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply docs/install/README.md]: Made first round of simple updates, assuming the cluster scope mode, to be consistent with Kserve. Will continue to refine.

File: config/rbac/role.yaml (hunk 1 above)
Comment (chinhuang007) [reply config/rbac/role.yaml]: Removed RBAC changes.

File: docs/install/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (chinhuang007) [reply docs/install/README.md:67]: Thanks for the clarification. If the session is about the resources created by code, not the user, the scenario for multi-namespace seems different. We actually create `model-serving-etcd` by copying one from the controller plane namespace and expect the user to create `storage-config`.

File: controllers/servingruntime_controller.go [hunk 9]
```diff
@@ -113,12 +115,77 @@ func (r *ServingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return RequeueResult, fmt.Errorf("could not reconcile the modelmesh type-constraints configmap: %w", err)
 	}
 
-	//TODO(probably) here ... if not controller namespace then
-	//  read etcd secret from controller namespace, replace rootprefix with ns-specific one
-	//  and the create/update etcd secret (with same name) in _this_ namespace
-	//  and include an "ownership" label similar to the tc-config configmap
+	// Reconcile the model mesh cluster config map and etcd secret
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply controllers/servingruntime_controller.go]: Good point. Done.

File: pkg/mmesh/etcd_secret.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	etcdEndpoints := strings.Split(es.EtcdConfig.Endpoints, ",")
+
+	newEtcdEndpoints := ""
+	// For each endpoint, insert controller namespace if not there
+	// TODO: revisit and come up with ideal solution
+	for i := range etcdEndpoints {
+		if !strings.Contains(etcdEndpoints[i], ".") {
+			// This looks like http://etcd:2379, so insert controller namespace before :port
+			parts := strings.Split(etcdEndpoints[i], ":")
+			etcdEndpoints[i] = parts[0] + ":" + parts[1] + "." + es.ControllerNamespace + ":" + parts[2]
```
[the remark this answers is not in the document]
Comment (njhill) [reply pkg/mmesh/etcd_secret.go:92]: Something like:
```go
re := regexp.MustCompile("(?:https?://)?([^\\s.:]+)(:\\d+)?")
if se := re.FindStringSubmatchIndex(etcdEndpoints[i]); se != nil {
    etcdEndpoints[i] = fmt.Sprintf("%s.%s%s", etcdEndpoints[i][:se[3]], es.ControllerNamespace, etcdEndpoints[i][se[3]:])
}
```

File: pkg/mmesh/modelmesh_events.go [hunk 11]
```diff
@@ -176,7 +176,7 @@ func (mes *ModelMeshEventStream) refreshWatches(nw *namespaceWatch, namespace, s
 			}
 			if owner, err := ownerIDFromVModelRecord(value); err == nil {
 				encodedNamespace := namespace
-				if owner != "" {
+				if owner != "" && (len(namespace) <= len(owner) || owner+"_" != namespace[:len(owner)+1]) {
```
Comment (njhill) [inline pkg/mmesh/modelmesh_events.go:179]: @chinhuang007 this can also be reverted I think... should be no changes at all needed to this file.

File: main.go [hunk 12]
```diff
@@ -190,7 +190,6 @@ func main() {
 	mgrOpts := ctrl.Options{
 		Scheme:                 scheme,
 		MetricsBindAddress:     metricsAddr,
-		Namespace:              ControllerNamespace,
```
[the remark this answers is not in the document]
Comment (chinhuang007) [reply main.go:193]: Env variable is added in the latest commit. Please take a look.

File: main.go [hunk 13]
```diff
@@ -187,10 +189,40 @@ func main() {
...
+	err = cl.Get(context.Background(), client.ObjectKey{Name: "foo"}, &corev1.Namespace{})
+	clusterScopeEnabled = err == nil || errors.IsNotFound(err)
+
+	if clusterScopeEnabled {
+		if namespaceScope != trueString {
+			setupLog.Info("Controller operating in cluster scope mode, will attempt to watch all namespaces")
+		} else {
+			// Config mismatch, namespace mode with cluster permissions, exit
+			setupLog.Info("In namespace scope mode but controller has cluster scope permissions, exit")
+			os.Exit(1)
```
Comment (njhill) [inline main.go:209]: I'm not sure we need to fail in this case, just have more permission than needed.

File: main.go [hunk 14]
```diff
@@ -187,10 +189,40 @@ func main() {
...
+		}
+	} else {
+		if namespaceScope != trueString {
+			// Config mismatch, cluster mode without cluster permissions, exit
+			setupLog.Info("In cluster scope mode but controller has namespace scope permissions, exit")
```
Comment (njhill) [inline main.go:214]: `Info` -> `Error`

File: pkg/mmesh/etcdrangewatcher.go [hunk 15]
```diff
@@ -30,17 +30,17 @@ import (
...
-	Password              string `json:"password"`
-	RootPrefix            string `json:"root_prefix"`
-	Certificate           string `json:"certificate"`
-	CertificateFile       string `json:"certificate_file"`
-	ClientKey             string `json:"client_key"`
-	ClientKeyFile         string `json:"client_key_file"`
-	ClientCertificate     string `json:"client_certificate"`
-	ClientCertificateFile string `json:"client_certificate_file"`
-	OverrideAuthority     string `json:"override_authority"`
+	Endpoints             string `json:"endpoints,omitempty"`
```
Comment (njhill) [inline pkg/mmesh/etcdrangewatcher.go:33]: Probably best not to have `omitempty` on this one since it's required (all the others are optional).

File: pkg/mmesh/etcd_secret.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		if i != 0 {
+			newEtcdEndpoints += ","
+		}
+		newEtcdEndpoints += etcdEndpoints[i]
+	}
+
+	es.EtcdConfig.Endpoints = newEtcdEndpoints
+	es.EtcdConfig.RootPrefix = fmt.Sprintf("%s/mm_ns/%s", es.EtcdConfig.RootPrefix, es.Namespace)
+
+	b, _ := json.Marshal(es.EtcdConfig)
```
Comment (njhill) [inline pkg/mmesh/etcd_secret.go:86]: Sorry forgot to add this in last review - should check for the error here and propagate it back

**Review by njhill — COMMENTED** [review]
@chinhuang007 just couple more very small things!

File: main.go (hunk 13 above)
Comment (chinhuang007) [reply main.go:209]: This might cause unexpected behaviors. The user would expect the controller to watch and reconcile in a single namespace but the controller is acting across all namespaces. I think we should exit and make sure the user/installer syncs up RBAC with install option. What do you think?

Comment (chinhuang007) [reply main.go:209]: Additionally clusterScopeEnabled will affect how service and SR controllers work internally. For ex, clusterScopeEnabled = true and namespaceScope = true would have different code path from clusterScopeEnabled = false and namespaceScope = true

Comment (njhill) [reply main.go:209]: @chinhuang007 OK sorry I think the thing I meant to say but forgot to add is that we should also change the later logic to depend on the namespaceScope setting instead of this RBAC check. Basically change the two places where `clusterScopeEnabled` is referenced lower down in this func. Are you ok with the above change if we also do that?

**Comment by chinhuang007** [comment]
@njhill I hope all your comments are addressed. Please take a look.

File: controllers/servingruntime_controller.go [hunk 17]
```diff
@@ -113,12 +115,60 @@ func (r *ServingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+		// replace rootprefix with ns-specific one, and then create/update etcd secret (with same name)
+		// in _this_ namespace and include labels similar to the tc-config configmap
+		s := &corev1.Secret{}
+		err = r.Client.Get(ctx, types.NamespacedName{
+			Name:      etcdSecretName,
+			Namespace: r.ControllerNamespace,
+		}, s)
+		if err != nil {
+			return RequeueResult, fmt.Errorf("Could not get the controller etcd secret: %w", err)
+		}
```
Comment (njhill) [inline controllers/servingruntime_controller.go:149]: nit:
```suggestion
		if err = r.Client.Get(ctx, types.NamespacedName{
			Name:      etcdSecretName,
			Namespace: r.ControllerNamespace,
		}, s); err != nil {
			return RequeueResult, fmt.Errorf("Could not get the controller etcd secret: %w", err)
		}
```

File: docs/install/README.md [hunk 18]
```diff
@@ -55,11 +60,12 @@ Please be aware that:
 
 For more details see the [built-in runtime configuration](../configuration/built-in-runtimes.md)
 
-In addition, the following resources will be created in the same namespace:
+The following resources will be created in the namespaces:
 
 - `model-serving-defaults` - ConfigMap holding default values tied to a release, should not be modified. Configuration can be overriden by creating a user ConfigMap, see [configuration](../configuration)
 - `tc-config` - ConfigMap used for some internal coordination
 - `storage-config` - Secret holding config for each of the storage backends from which models can be loaded - see [the example](../predictors/)
+- `model-serving-etcd` - Secret providing access to the Etcd cluster. It is created by user in the controller namespace - see [instructions](../install/install-script.md#setup-the-etcd-connection-information), and will be automatically created in user namespaces in the cluster scope mode.
```
Comment (njhill) [inline docs/install/README.md:68]: small suggested wording tweak:
```suggestion
- `model-serving-etcd` - Secret providing access to the Etcd cluster. It is created by user in the controller namespace - see [instructions](../install/install-script.md#setup-the-etcd-connection-information), and will be automatically created in user namespaces when in cluster scope mode.
```

File: pkg/mmesh/etcd_secret.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	EtcdConnectionKey = "etcd_connection"
+)
```
Comment (njhill) [inline pkg/mmesh/etcd_secret.go:33]: Can use existing constant `modelmesh.EtcdConnectionKey` for this

File: pkg/mmesh/etcd_secret.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+		},
+	}
+	err := es.addData(s)
+	if err != nil {
+		return err
+	}
+	err = cl.Create(ctx, s)
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = cl.Update(ctx, s)
+	}
```
Comment (njhill) [inline pkg/mmesh/etcd_secret.go:64]: The typical state will be already created, so this will cause some unnecessary churn... would be best to do a Get first above before adding the data (which will come from local cache), and then create only if it doesn't exist. If it _does_ exist then just update (or even better check first to see if the data already matches in which case there's nothing to do)

**Review by njhill — COMMENTED** [review]
Hey @chinhuang007 looks great thanks, just couple of remaining comments (sorry!), hopefully should be pretty quick.

**Comment by chinhuang007** [comment]
@njhill thanks for the review. hopefully addressed your comments.

**Review by njhill — APPROVED** [review]
Thanks @chinhuang007 for all your work on this and your patience!

**Review by pvaneck — APPROVED** [review]
Thanks, @chinhuang007 and @njhill!
/lgtm
